pkg/controller/grouping/group_entity_index.go

```diff
@@ -0,0 +1,818 @@
+	}
+}
+
+func (i *GroupEntityIndex) Run(stopCh <-chan struct{}) {
+	klog.Info("Starting GroupEntityIndex")
+	for {
+		select {
+		case <-stopCh:
+			klog.Info("Stopping GroupEntityIndex")
+			return
+		case group := <-i.eventChan:
+			parts := strings.Split(group, "/")
+			groupType, name := GroupType(parts[0]), parts[1]
+			for _, handler := range i.eventHandlers[groupType] {
+				handler(name)
+			}
+		}
+	}
+}
+
+func (i *GroupEntityIndex) AddEventHandler(groupType GroupType, handler eventHandler) {
+	i.eventHandlers[groupType] = append(i.eventHandlers[groupType], handler)
+}
+
+func (i *GroupEntityIndex) HasSynced() bool {
+	i.lock.RLock()
+	defer i.lock.RUnlock()
+	if i.initialPodCount == nil || i.maxPodCount < *i.initialPodCount {
+		return false
+	}
+	if i.initialExternalEntityCount == nil || i.maxExternalEntityCount < *i.initialExternalEntityCount {
+		return false
+	}
+	if i.initialNamespaceCount == nil || i.maxNamespaceCount < *i.initialNamespaceCount {
+		return false
+	}
+	return true
+}
+
+func (i *GroupEntityIndex) setInitialCounts(initialNamespaceCount, initialPodCount, initialExternalEntityCount int) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+	i.initialNamespaceCount = &initialNamespaceCount
+	i.initialPodCount = &initialPodCount
+	i.initialExternalEntityCount = &initialExternalEntityCount
+}
+
+func (i *GroupEntityIndex) match(entityType entityType, label labels.Set, namespace string, sel *types.GroupSelector) bool {
+	objSelector := sel.PodSelector
+	if entityType == externalEntityType {
+		objSelector = sel.ExternalEntitySelector
+	}
+	if sel.Namespace != "" {
+		if sel.Namespace != namespace {
+			// Pods or ExternalEntities must be matched within the same Namespace.
+			return false
+		}
+		if objSelector != nil && !objSelector.Matches(label) {
+			// podSelector or externalEntitySelector exists but doesn't match the ExternalEntity or Pod's labels.
+			return false
+		}
+		return true
+	}
+	if sel.NamespaceSelector != nil {
+		if !sel.NamespaceSelector.Empty() {
+			namespaceLabels, exists := i.namespaceLabels[namespace]
+			if !exists {
+				return false
+			}
+			if !sel.NamespaceSelector.Matches(namespaceLabels) {
+				// Pod's Namespace do not match namespaceSelector.
+				return false
+			}
+		}
+		if objSelector != nil && !objSelector.Matches(label) {
+			// ExternalEntity or Pod's Namespace matches namespaceSelector but
+			// labels do not match the podSelector or externalEntitySelector.
+			return false
+		}
+		return true
+	}
+	if objSelector != nil {
+		// Selector only has a PodSelector/ExternalEntitySelector and no sel.Namespace.
+		// Pods/ExternalEntities must be matched from all Namespaces.
+		if !objSelector.Matches(label) {
+			// pod/ee labels do not match PodSelector/ExternalEntitySelector.
+			return false
+		}
+		return true
+	}
+	// TODO: if none of the above selector is set, it means matching all entities or none?
+	// This is only possible for ClusterGroup. From previous code, it means matching none.
```

Comment:
Sorry for resuscitating that conversation, but I feel like the comment should make it clear that "unset" here is either missing (`nil`) or empty. Unless I misunderstood the thread.

It seem that this doesn't need to be a TODO either - this is just the behavior we have and it seems we do not intend to change it.

Reply:
removed TODO and made it more clear.